File: BatchRenamer/FileRenamer.cs
using System;
using System.IO;
using System.Linq;
using System.Text.RegularExpressions;

namespace BatchRenamer
{
    public static class FileRenamer
    {
        public delegate string Renamer(string original, ReplaceOptions options);
        private static RegexUtils regexUtils = new RegexUtils();

        public static string StringReplace(string original, ReplaceOptions options)
        {
            string findEscaped = Regex.Escape(options.Find);
            string replaceEscaped = Regex.Escape(options.Replace);
            RegexOptions regexOptions = options.IgnoreCase ? RegexOptions.IgnoreCase : RegexOptions.None;
            string replaced = Regex.Replace(original, findEscaped, replaceEscaped, regexOptions);
            return Regex.Unescape(replaced);
        }

        public static string RegexReplace(string original, ReplaceOptions options)
        {
            RegexOptions regexOptions = options.IgnoreCase ? RegexOptions.IgnoreCase : RegexOptions.None;
            // cache last failed pattern to prevent excessive exceptions
            return regexUtils.CachedTryReplace(original, options.Find, options.Replace, regexOptions, out bool success);
        }

        public static string RenamePath(string path, ReplaceOptions options, Renamer renamer, bool renameExtension)
        {
            string directory = Path.GetDirectoryName(path);
            string fileName = Path.GetFileName(path);
            return $"{directory}{Path.DirectorySeparatorChar}{RenameFile(fileName, options, renamer, renameExtension)}";
            // can't use Path.Combine, because it throws an exception when fileName is invalid
        }

        public static string RenameFile(string fileName, ReplaceOptions options, Renamer renamer, bool renameExtension)
        {
            string name = Path.GetFileNameWithoutExtension(fileName);
            string extension = Path.GetExtension(fileName);

            return renameExtension ?
                renamer(name + extension, options) :
                renamer(name, options) + extension;
        }

        public static string DisplayName(string path, bool showExtension)
        {
            return showExtension ?
                Path.GetFileName(path) :
                Path.GetFileNameWithoutExtension(path);
        }

        public static bool IsInvalidFileName(string fileName)
        {
            return Path.GetInvalidFileNameChars()
                .Any(c => fileName.Contains(c));
        }
    }
}

File: Renaming/FileRenamer.fs
namespace Renaming
module FileRenamer =
    open System.Text.RegularExpressions
    open System.IO

    type ReplaceOptions = { Find: string; Replace: string; IgnoreCase: bool }
    type Renamer = ReplaceOptions -> string -> string

    let stringReplace { Find = findStr; Replace = replaceStr; IgnoreCase = ignoreCase } (fileName: string) =
        let findEscaped = Regex.Escape(findStr)
        let replaceEscaped = Regex.Escape(replaceStr)
        let regexOptions = if ignoreCase then RegexOptions.IgnoreCase else RegexOptions.None
        let replaced = Regex.Replace(fileName, findEscaped, replaceEscaped, regexOptions)
        Regex.Unescape(replaced)

    let regexReplace { Find = findRegex; Replace = replaceRegex; IgnoreCase = ignoreCase } (fileName: string) =
        let regexOptions = if ignoreCase then RegexOptions.IgnoreCase else RegexOptions.None
        Regex.Replace(fileName, findRegex, replaceRegex, regexOptions)

    type FileInfo = {Directory: string; Name: string; Extension: string}
    
    let fileInfo (path: string) =
        { Directory = Path.GetDirectoryName(path);
          Name = Path.GetFileNameWithoutExtension(path);
          Extension = Path.GetExtension(path) }

    let renamedFile renamer options renameExtension path =
        let file = fileInfo path
        let renamedFile =
            if renameExtension
            then renamer options (file.Name + file.Extension)
            else (renamer options file.Name) + file.Extension
        Path.Combine(file.Directory, renamedFile)

    let renamedFiles renamer files =
        Seq.map (fun original -> original, renamer original) files

    let displayName showExtension (path: string) =
        if showExtension
        then Path.GetFileName path
        else Path.GetFileNameWithoutExtension path

    let isInvalid (path: string) =
        let invalidFileNameChars = Path.GetInvalidFileNameChars() |> Seq.map string
        let invalidPathChars = Path.GetInvalidPathChars() |> Seq.map string
        let path = Path.GetDirectoryName(path)
        let fileName = Path.GetFileName(path)
        (Seq.exists (fun (invalid: string) -> fileName.Contains(invalid)) invalidFileNameChars)
        || (Seq.exists (fun (invalid: string) -> path.Contains(invalid)) invalidPathChars)

    type ValidationResult =
        | Valid=1
        | IsDuplicate=2
        | InvalidName=4

    let validate names file =
        if Seq.contains file names then ValidationResult.IsDuplicate
        else if isInvalid file then ValidationResult.InvalidName
        else ValidationResult.Valid
